### ISSUE TITLE
Integration with next-auth fails, routing issue

### ISSUE THREAD

**Issue by neophob** [fastify/fastify-nextjs#180]
## 🐛 Bug Report

I want to integrate next-auth with my fastify-nextjs project. This worked until the next.auth library does the login/POST callback - somehow this call does not succeed and just times out. I ported the exact same code to a vanilla Next.js project (using `npx create-next-app`) and there it works flawlessly!

## To Reproduce

- Fastify Next.js example which does not work: https://github.com/neophob/nextjs-fastify-auth
- Next.js only example which does work: https://github.com/neophob/nextjs-auth

## Expected behavior

call is terminated correct / callback is called. I quickly added some console log into the plugin to figure out where the code hangs:

```
    function handler (req, reply) {
      if (callback) {
        return callback(app, req, reply)
      }
      console.log("handle next request", req.params)
      return handleNextRequests(req.raw, reply.raw)
        .then(() => {
          console.log("...handled")
          reply.sent = true
        })
```

- when the login POST request is hanging, the `handleNextRequests` function is running
- when I cancel the POST request (using postman), then the `...handled` console output is visible...

the console shows this output:

```
{"level":30,"time":1613721952594,"pid":94353,"hostname":"hostname","reqId":29,"req":{"method":"POST","url":"/api/auth/callback/credentials","hostname":"localhost:3333","remoteAddress":"::1","remotePort":55873},"msg":"incoming request"}
handle next request { '*': 'auth/callback/credentials' }
event - build page: /api/auth/[...nextauth]
wait  - compiling...
event - compiled successfully
```

To summarize, I think the POST call to `/api/auth/callback/credentials` is either not routed correctly OR the return value is invalid (promise vs. value maybe?)

## Your Environment

- *node version*: 12 and 14
- *fastify version*: 3.12
- *os*: Mac

**Comment by mcollina** [comment]
As you can see in https://github.com/fastify/fastify-nextjs/blob/2f2f9e0195c29db78a0563cc0b12984af8cc6b6a/index.js#L48, next is only exposed via GETs. You can call `fastify.next(path, 'post')` to also expose POSTs. Or maybe we can use `'all'` to publish it to all HTTP methods.

**Comment by neophob** [comment]
Jep I saw that, thats why I did the PR to update the docs ;) I think I handle this correctly in my code where I do:

```
  fastify
    .register(fastifyNextJs, { dev: true, logLevel: 'debug' })
    .after(() => {
      fastify.next('/api/*', { method: 'get' });
      fastify.next('/api/*', { method: 'post' });
    });
```

**Comment by mcollina** [comment]
Weird, I thought it should work. WOuld you like to send a PR to fix this?

**Comment by neophob** [comment]
Yeah, I tried to debug this issue, but somehow i failed ro fix this. do you have a hint for me?

**Comment by mcollina** [comment]
I _think_ the POST body might already be parsed when it reaches next.

**Comment by neophob** [comment]
Hmm even if we use `handleNextRequests(req.raw, reply.raw)`?

what happen here is the `handleNextRequests` call never returns:
```
      return handleNextRequests(req.raw, reply.raw)
        .then(() => {
```

when the request is canceled THEN `handleNextRequests` resolves. So I replaced the `const handleNextRequests = app.getRequestHandler()` with `const handleNextRequests = app.router.execute.bind(app.router)` as a quick hack - with the same result as before. But this means there's an error with the routing part...

Other suspects: `'content-type': 'application/x-www-form-urlencoded',` - I could not really find docs which say `x-www-form-urlencoded` is supported or not, but after reading https://github.com/vercel/next.js/discussions/12542 I'm not sure anymore...

**Comment by mcollina** [comment]
It needs to be investigated but likely doable.

**Comment by simoneb** [comment]
@neophob can you provide a repro for this issue?

**Comment by neophob** [comment]
@simoneb sure, I've linked them in the original report:

- Fastify Next.js example which does not work: https://github.com/neophob/nextjs-fastify-auth
- Next.js only example which does work: https://github.com/neophob/nextjs-auth

Thanks!

**Comment by simoneb** [comment]
@neophob this can be easily fixed in userland by removing the `fastify-formbody` plugin registration and providing a dummy passthrough body parser. See the attached patch. [0001-passthough-body-parser.txt](https://github.com/fastify/fastify-nextjs/files/6068863/0001-passthough-body-parser.txt)

I'm not sure if this requires a more generic implementation to be included in this library though.

**Comment by neophob** [comment]
Thanks @simoneb for your help! I applied the patch to my repo at https://github.com/neophob/nextjs-fastify-auth. Looks like the data is now passed down further but somehow the data did not arrive at nextjs now:

```
{"level":30,"time":1614717014846,"pid":6340,"hostname":"laptop","reqId":22,"req":{"method":"POST","url":"/api/auth/callback/credentials","hostname":"localhost:3333","remoteAddress":"::1","remotePort":51064},"msg":"incoming request"}
{"level":50,"time":1614717014847,"pid":6340,"hostname":"laptop","reqId":22,"req":{"method":"POST","url":"/api/auth/callback/credentials","hostname":"localhost:3333","remoteAddress":"::1","remotePort":51064},"res":{"statusCode":500},"err":{"type":"TypeError","message":"Converting circular structure to JSON\n    --> starting at object with constructor 'Socket'\n    |     property 'parser' -> object with constructor 'HTTPParser'\n    --- property 'socket' closes the circle","stack":"TypeError: Converting circular structure to JSON\n    --> starting at object with constructor 'Socket'\n    |     property 'parser' -> object with constructor 'HTTPParser'\n    --- property 'socket' closes the circle\n    at JSON.stringify (<anonymous>)\n    at Object.handler (/nextjs-fastify-auth/node_modules/fastify-nextjs/index.js:62:97)\n    at preHandlerCallback (/nextjs-fastify-auth/node_modules/fastify/lib/handleRequest.js:124:28)\n    at preValidationCallback (/nextjs-fastify-auth/node_modules/fastify/lib/handleRequest.js:107:5)\n    at handler (/nextjs-fastify-auth/node_modules/fastify/lib/handleRequest.js:70:7)\n    at done (/nextjs-fastify-auth/node_modules/fastify/lib/contentTypeParser.js:128:7)\n    at /nextjs-fastify-auth/node_modules/fastify/lib/contentTypeParser.js:119:27"},"msg":"Converting circular structure to JSON\n    --> starting at object with constructor 'Socket'\n    |     property 'parser' -> object with constructor 'HTTPParser'\n    --- property 'socket' closes the circle"}
{"level":30,"time":1614717014848,"pid":6340,"hostname":"laptop","reqId":22,"res":{"statusCode":500},"responseTime":1.9242780208587646,"msg":"request completed"}
```

**Comment by neophob** [comment]
@simoneb awesome, thanks - I tested it and it worked! But I agree about the core question, how should this "fix" be applied, updated documentation in the `fastify-nextjs` package or is the core issue more in the fastify core? Because when I hit this issue and figured out "oh - my app needs to parse `application/x-www-form-urlencoded`" - so obvious I installed the package `fastify-formbody`...

**Comment by simoneb** [comment]
It's a good question and didn't think thoroughly about this. The thing is:

- you need to get the request body to nextjs, and that has to go through fastify
- therefore fastify needs to be able  to accept that request body with that content type
- but fastify should not consume the request body because it's intended to be consumed by nextjs

Therefore there needs to be a passthough done at some point. I guess that could be done by this plugin so that whenever a nextjs route is registered, a catch-all body parser is added to that route which passes the raw body to nextjs. 

This sounds viable to me but I would like to hear @mcollina's thoughts too.

**Comment by mcollina** [comment]
I'm +1 with the pass through content parser.

**Comment by simoneb** [comment]
@neophob would you like to submit a PR to include this feature in this plugin?

**Comment by neophob** [comment]
Sure - but tbh I'm not 100% sure we have the same understanding. To summarize my understanding:

- the pass through solution (as demonstrated in your PR) is fine
- to improve developer experience we document the finding of this very issue, the pr will update the readme

does that make sense?

**Comment by simoneb** [comment]
Not exactly, we would like to include this feature in the plugin so it doesn't have to be done in user code.

**Comment by neophob** [comment]
Oh I see. Implementing the type parser in this plugin would make it easier to use (less options to configure) - this would be a plus. But if I take a step back and see were I want to use my Next.JS app: it's an existing backend application with a frontend. So what if my backend already needs `fastify-formbody`? Then this solution won't work anymore I assume and we are back at the start and need to provide more documentation.

Wouldn't a "redirect under this path to a plugin" fastify option make more sense? For example this plugin would be installed with an optional parameter for the routing like this:

```
   fastify
     .register(fastifyNextJs, { path: '/next' } ,{ dev: true, logLevel: 'debug' })
```
then everything below that path would automatically be routed. O

**Comment by simoneb** [comment]
I like the idea but it will change the semantics of how the plugin currently works, which is via server decorators. I'm not sure why the plugin was designed this way, I guess for good reasons, but I like the idea of using a plugin registration option to handle this.

**Comment by mcollina** [comment]
I think we can remove the decorator if we want to, there is no need for it. It'd be a semver major change but likely better.

**Comment by simoneb** [comment]
@neophob are you considering to send a PR to address this or are you happy with the fix in user-land?

**Comment by neophob** [comment]
The user-land fix works for me! Thanks again!
